medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,590 @@
+        )
+        time_window_attribute = self._get_attribute_name(medrecord, "time_window")
+        number_of_windows_attribute = self._get_attribute_name(
+            medrecord, "number_of_windows"
+        )
+
+        # TODO: need to perfect this with new queries.
+        for patient_index in sorted(medrecord.nodes_in_group(self.patients_group)):
+            first_admission = medrecord.node[patient_index, first_admission_attribute]
+            if not isinstance(first_admission, datetime.datetime):
+                raise ValueError(
+                    f"First admission attribute needs to be a datetime object, but got {type(first_admission)}"
+                )
+            for single_edge in medrecord.edges_connecting(
+                patient_index,
+                sorted(medrecord.nodes_in_group(self.concepts_group)),
+                directed=False,
+            ):
+                time = medrecord.edge[single_edge, self.time_attribute]
+                if not isinstance(time, datetime.datetime):
+                    raise ValueError(
+                        f"Time attribute needs to be a datetime object, but got {type(time)}"
+                    )
+                medrecord.edge[single_edge, absolute_time_window_attribute] = floor(
+                    (time - first_admission).days // time_interval_days
+                )
+                medrecord.edge[single_edge, concept_index_attribute] = (
+                    medrecord.select_nodes(
+                        node()
+                        .index()
+                        .is_in(list(medrecord.edge_endpoints(single_edge)))
+                        & node().in_group(self.concepts_group)
+                    )[0]
+                )
+
+            time_windows = self._remove_low_number_time_windows(
+                medrecord,
+                patient_index,
+                absolute_time_window_attribute,
+                min_codes_per_window,
+            )
+            self._assign_relative_time_windows(
+                medrecord=medrecord,
+                patient_index=patient_index,
+                time_windows=time_windows,
+                absolute_time_window_attribute=absolute_time_window_attribute,
+                time_window_attribute=time_window_attribute,
+                number_of_windows_attribute=number_of_windows_attribute,
+            )
+
+        preprocessing_attributes = PreprocessingAttributes(
+            first_admission_attribute=first_admission_attribute,
+            time_window_attribute=time_window_attribute,
+            concept_index_attribute=concept_index_attribute,
+            number_of_windows_attribute=number_of_windows_attribute,
+            absolute_time_window_attribute=absolute_time_window_attribute,
+        )
+
+        return preprocessing_attributes
+
+    def _assign_concept_indices(
+        self, medrecord: MedRecord, concept_index_attribute: MedRecordAttribute
+    ) -> Dict[int, NodeIndex]:
```

Comment:
Should probably be a List[NodeIndex] and the index of the concept is the index in the list. You could also create a type alias called `Concept` to make it clearer that it is a list of concepts